Backends/CLX/input.lisp

```diff
@@ -122,230 +120,214 @@
 ;;; manager is responsible for all setting of input focus for us.  If
 ;;; we want to do something more complicated with server input focus,
 ;;; then this method should be adjusted appropriately and the
 ;;; top-level-sheet REALIZE-MIRROR method should be adjusted to add
 ;;; :WM_TAKE_FOCUS to XLIB:WM-PROTOCOLS.  CSR, 2009-02-18
-(defmethod port-wm-protocols-message (sheet time (message (eql :wm_take_focus)) data)
-  (let ((timestamp (elt data 1))
-        (mirror (sheet-xmirror sheet)))
-    (when mirror
-      (xlib:set-input-focus (clx-port-display *clx-port*)
-                            mirror :parent timestamp))
-    nil))
-
-(defmethod port-wm-protocols-message (sheet time (message (eql :wm_delete_window)) data)
-  (declare (ignore data))
-  (make-instance 'window-manager-delete-event :sheet sheet :timestamp time))
-
-(defmethod port-wm-protocols-message (sheet time (message t) data)
-  (warn "Unprocessed WM Protocols message: ~:_message = ~S;~:_ data = ~S;~_ sheet = ~S."
-        message data sheet))
-
-
-
-(defmethod get-next-event ((port clx-basic-port) &key wait-function (timeout nil))
-  (declare (ignore wait-function))
-  (let* ((*clx-port* port)
-         (display    (clx-port-display port)))
-    (unless (xlib:event-listen display)
-      (xlib:display-force-output (clx-port-display port)))
-    ; temporary solution
-    (or (xlib:process-event (clx-port-display port) :timeout timeout :handler #'event-handler :discard-p t)
-	:timeout)))
-;; [Mike] Timeout and wait-functions are both implementation
-;;        specific and hence best done in the backends.
 
+(defun port-client-message (sheet time type data)
+  (case type
+    (:wm_protocols
+     (let ((message (xlib:atom-name (slot-value *clx-port* 'display) (aref data 0))))
+       (case message
+         (:wm_take_focus
+          (when-let ((mirror (sheet-xmirror sheet)))
+            (xlib:set-input-focus (clx-port-display *clx-port*)
+                                  mirror :parent (elt data 1)))
+          nil)
+         (:wm_delete_window
+          (make-instance 'window-manager-delete-event :sheet sheet :timestamp time))
+         (otherwise
+          (warn "Unprocessed WM Protocols message: ~:_message = ~S;~:_ data = ~S;~_ sheet = ~S."
+                message data sheet)))))
+    (otherwise
+     (warn "Unprocessed client message: ~:_type = ~S;~:_ data = ~S;~_ sheet = ~S."
+           type data sheet))))
+
+(defmethod process-next-event ((port clx-basic-port) &key wait-function (timeout nil))
+  (let ((*clx-port* port)
+        (*wait-function* wait-function))
+    (let ((event (xlib:process-event (clx-port-display port)
+                                     :timeout timeout
+				     :handler #'event-handler
+                                     :discard-p t
+                                     :force-output-p t)))
+      (case event
+        ((nil) (values nil :timeout))
+        ((t)   (values nil :wait-function))
```

Comment:
@scymtym maybe this one has confused you with regard to event-listen-or-wait. this interface behaves differently, because it is meant to return an event, while event-*-listen faimly returns either T (if some condition is met) or NIL (if it is not met, i.e due to timeout).